src/Common/src/Interop/Unix/libcurl/Interop.SafeCurlHandle.cs

```diff
@@ -56,11 +96,175 @@ public static void DisposeAndClearHandle(ref SafeCurlMultiHandle curlHandle)
+                }
+
+                unsafe
+                {
+                    // Write to special fd
+                    byte* dummyBytes = stackalloc byte[1];
+                    if ((int)libc.write(_specialFds[libc.WriteEndOfPipe], dummyBytes, (size_t)1) <= 0)
+                    {
+                        // TODO: How to handle errors?
+                        throw new InvalidOperationException("Cannot write data: " + Marshal.GetLastWin32Error());
+                    }
+                }
+            }
+
+            protected override void Dispose(bool disposing)
+            {
+                if (disposing)
+                {
+                    if (null != _timer)
+                    {
+                        _timer.Dispose();
+                    }
+                }
+                base.Dispose(disposing);
+            }
+
             protected override bool ReleaseHandle()
             {
-                Interop.libcurl.curl_multi_cleanup(this.handle);
+                Debug.Assert(0 == _fdSet.Count);
+                Debug.Assert(0 == _fdCount);
+                Debug.Assert(_pollCancelled);
+
+                libc.close(_specialFds[libc.ReadEndOfPipe]);
+                libc.close(_specialFds[libc.WriteEndOfPipe]);
+                libcurl.curl_multi_cleanup(this.handle);
+
                 return true;
             }
+
+            private int ReadSpecialFd(PollFlags revents)
+            {
+                PollFlags badEvents = PollFlags.POLLERR | PollFlags.POLLHUP | PollFlags.POLLNVAL;
+                if ((revents & badEvents) != 0)
+                {
+                    return -1;
+                }
+                int pipeReadFd = _specialFds[libc.ReadEndOfPipe];
+                int bytesRead = 0;
+                unsafe
+                {
+                    do
+                    {
+                        int bufferLength = 1024;
+                        byte* dummyBytes = stackalloc byte[bufferLength];
+                        int numBytes = (int)libc.read(pipeReadFd, dummyBytes, (size_t)bufferLength);
+                        if (numBytes > 0)
+                        {
+                            PollFlags outFlags;
+                            int retVal = libc.poll(pipeReadFd, PollFlags.POLLIN, 0, out outFlags);
+                            if (retVal < 0)
+                            {
+                                numBytes = -1;
+                            }
+                            else
+                            {
+                                bytesRead += numBytes;
+                                if (0 == retVal)
+                                {
+                                    break;
+                                }
+                            }
+                        }
+                        if (numBytes <= 0)
+                        {
+                            return -1;
+                        }
+                    }
+                    while (true);
```

Comment:
FWIW, I'm finding all of these conditionals hard to follow.  I'd personally find something like following easier to understand:
```C#
int bufferLength = 1024;
byte* dummyBytes = stackalloc byte[bufferLength];
int bytesRead = 0;
while (true)
{
    // We know there's data on the descriptor: read it.
    int numBytes = (int)libc.read(pipeReadFd, dummyBytes, (size_t)bufferLength);
    if (numBytes <= 0)
        return -1;
    bytesRead += numBytes;
    
    // See if there's any more data to be read
    PollFlags ignoredFlags;
    int retVal = libc.poll(pipeReadFd, PollFlags.POLLIN, 0, out ignoredFlags);
    if (retVal == 0)
        break;
    if (retVal < 0)
        return -1;
}
return bytesRead;
```